dist/IO/lib/IO/Socket.pm

```diff
@@ -396,218 +396,468 @@ IO::Socket - Object interface to socket communications
+Get an option associated with the socket. Levels other than C<SOL_SOCKET>
+may be specified here. As a convenience, this method will unpack a byte buffer
+of the correct size back into a number.
 
-=item *
+=head2 listen
 
-C<FLAGS> is optional and defaults to C<0>, and
+    $sock->listen(5);
 
-=item *
+Does the same thing that the C<listen(2)> system call does. Returns true if it
+succeeded, false otherwise. Listens to a socket with a given queue size.
 
-after a successful send with C<TO>, further calls to send() on an
-unconnected socket without C<TO> will send to the same address, and
-C<TO> will be used as the result of peername().
+=head2 peername
 
-=back
+    my $sockaddr_in = $sock->peername();
 
-=item recv(BUF, LEN, [,FLAGS])
+Returns the packed C<sockaddr> address of the other end of the socket
+connection. It calls C<getpeername>.
 
-Like the built-in L<recv()|perlfunc/recv>, except that:
 
-=over
+=head2 protocol
 
-=item *
+    my $proto = $sock->protocol();
 
-C<FLAGS> is optional and defaults to C<0>, and
+Returns the number for the protocol being used on the socket, if
+known. If the protocol is unknown, as with an C<AF_UNIX> socket, zero
+is returned.
 
-=item *
+=head2 recv
 
-the cached value returned by peername() is updated with the result of
-recv().
+    my $buffer = "";
+    my $length = 1024;
+    my $flags = 0; # default. optional
+    $sock->recv($buffer, $length);
+    $sock->recv($buffer, $length, $flags);
 
-=back
+Similar in functionality to L<perlfunc/recv>.
 
-=item peername
+Receives a message on a socket. Attempts to receive C<$length> characters of
+data into C<$buffer> from the specified socket. C<$buffer> will be grown or
+shrunk to the length actually read. Takes the same flags as the system call of
+the same name. Returns the address of the sender if socket's protocol supports
+this; returns an empty string otherwise. If there's an error, returns
+C<undef>. This call is actually implemented in terms of the C<recvfrom(2)>
+system call.
 
-Returns the cached peername, possibly set by recv() or send() above.
-If not otherwise set returns (and caches) the result of getpeername().
+Flags are ORed together values, such as C<MSG_BCAST>, C<MSG_OOB>,
+C<MSG_TRUNC>. The default value for the flags is C<0>.
 
-=item protocol
+The cached value of L<IO::Socket/"peername"> is updated with the result of
+C<recv>.
 
-Returns the numerical number for the protocol being used on the socket, if
-known. If the protocol is unknown, as with an AF_UNIX socket, zero
-is returned.
+Note the characters: depending on the status of the socket, either (8-bit)
+bytes or characters are received. By default all sockets operate on bytes, but
+for example if the socket has been changed using C<binmode> to operate with
+the C<:encoding(UTF-8)> I/O layer (see the C<open> pragma), the I/O will
+operate on UTF8-encoded Unicode characters, not bytes. Similarly for the
+C<:encoding> layer: in that case pretty much any characters can be read.
```

Comment:
This is no longer true.

````
$ ./perl -Ilib -MIO::Socket -e '$s = IO::Socket->new(Domain => AF_INET, Type => SOCK_STREAM, proto => "tcp", PeerHost => "127.0.0.1", PeerPort => 80) or die; binmode $s, ":utf8"; $s->send("foo")'
send() isn't allowed on :utf8 handles at lib/IO/Socket.pm line 295.
````

Reply:
Would you suggest removing that paragraph entirely?

Reply:
I think that is best. It seems to be from the recv docs in perlfunc, so the new warning about `:utf8` could be copied from there.

Reply:
`send` should also probably have the same warning for consistency.

Reply:
This is complicated by this being dual-life.  Older perls will behave as the block I highlighted described, new perls will throw the error.

Reply:
True, I guess it needs to reference both behaviors depending on version.

Reply:
Would the following work well enough for both cases?

```POD
B<Note:> In Perl v5.30 and newer, if the socket has been marked as C<:utf8>,
C<recv> will throw an exception. The C<:encoding(...)> layer implicitly
introduces the C<:utf8> layer. See L<perlfunc/binmode>.

B<Note:> In Perl versions older than v5.30, depending on the status of the
socket, either (8-bit) bytes or characters are received. By default all
sockets operate on bytes, but for example if the socket has been changed
using L<perlfunc/binmode> to operate with the C<:encoding(UTF-8)> I/O layer
(see the L<perlfunc/open> pragma), the I/O will operate on UTF8-encoded
Unicode characters, not bytes. Similarly for the C<:encoding> layer: in
that case pretty much any characters can be read.
```